src/Http/Http.Extensions/src/DefaultProblemDetailsWriter.cs

```diff
@@ -59,26 +65,43 @@ public ValueTask WriteAsync(ProblemDetailsContext context)
         ProblemDetailsDefaults.Apply(context.ProblemDetails, httpContext.Response.StatusCode);
         _options.CustomizeProblemDetails?.Invoke(context);
 
-        // Use source generation serialization in two scenarios:
-        // 1. There are no extensions. Source generation is faster and works well with trimming.
-        // 2. Native AOT. In this case only the data types specified on ProblemDetailsJsonContext will work.
-        if (context.ProblemDetails.Extensions is { Count: 0 } || !RuntimeFeature.IsDynamicCodeSupported)
+        var problemDetailsType = context.ProblemDetails.GetType();
+
+        if (problemDetailsType == typeof(ProblemDetails) ||
+            problemDetailsType == typeof(HttpValidationProblemDetails))
         {
-            return new ValueTask(httpContext.Response.WriteAsJsonAsync(
-                context.ProblemDetails,
-                ProblemDetailsJsonContext.Default.ProblemDetails,
-                contentType: "application/problem+json"));
+            return WriteKnownProblemDetailsAsync(context);
         }
 
         return new ValueTask(httpContext.Response.WriteAsJsonAsync(
                         context.ProblemDetails,
-                        options: null,
+                        _serializerOptions.GetTypeInfo(problemDetailsType),
                         contentType: "application/problem+json"));
+
+        ValueTask WriteKnownProblemDetailsAsync(ProblemDetailsContext context)
+        {
+            // Use source generation serialization in two scenarios:
+            // 1. There are no extensions. Source generation is faster and works well with trimming.
```

Comment:
Not a fan using the internal context but, since converters are internal, I will keep as is until PD AOT support is better discussed.